### PR TITLE
Optimize calls to .cpu() during checkpointing

### PR DESCRIPTION
**Pull request by ksivaman**
This is a better way to solving the problem in #351 as we have since observed that the number of `.cpu()` calls during saving and loading the checkpoint results in a non-negligible overhead. Just for the state tensors, there are 18 calls per transformer layer and this is excluding `.cpu()` calls to the global fp8 buffer.

### CODE REVIEW

**Review by timmoon10 — APPROVED** [review]
LGTM

File: transformer_engine/pytorch/module/base.py [hunk 1]
```diff
@@ -409,8 +409,12 @@ def set_extra_state(self, state: torch.Tensor) -> None:
 
         if isinstance(state, torch.Tensor):
             state = pickle.loads(state.detach().cpu().numpy().tobytes())
-            if state is None:
-                return
+        elif isinstance(state, io.BytesIO):
+            state.seek(0)
+            state = torch.load(io.BytesIO(state.read()), map_location='cuda')
```
Comment (mikolajblaz) [inline transformer_engine/pytorch/module/base.py:414]: Could it be just
```
state = torch.load(state, map_location='cuda')
```
?

**Comment by nzmora-nvidia** [comment]
LGTM.
Next opportunity I'll add a remark explaining why we're doing that strange serialization to a torch.tensor for ONNX so it doesn't confuse someone else later.

**Pull request merged**